Plugin development draft PR (depreciated)

woensug-choi opened the pull request:
Direct Kinematics plugin that connects between `/gazebo/set_model_state` and `frl_vehicle_msg/UwGliderCommand` is ready.

There are much to be improved, but it may do good enough if we are to test the initial connection with the control system/backseat driver

### Features and How-to
- Need frl_vehicle_msg installed
```bash
cd ~/uuv_ws/src
git clone git@github.com:Field-Robotics-Lab/frl_msgs.git
cd ..
catkin_make
```
- The glider and its initial position is defined at `/glider_hybrid_whoi_description/urdf/glider_hybrid_whoi_base_kinematics.xacro`
- You may run with followings
```bash
roslaunch glider_hybrid_whoi_gazebo empty_underwater_world.launch paused:=false
roslaunch glider_hybrid_whoi_gazebo start_demo_teleop_kinematics.launch joy_id:=0
```
- You may run the test python script with the following command
```bash
rosrun direct_kinematics_ros_plugins test_directKinematics.py
```
- You can edit the specification and numbers of commands at `glider_hybrid_whoi/direct_kinematics_ros_plugins/test/test_directKinematics.py`

### Test Results
___Thruster function is not implemented here, nor correct properties of glider models and their dynamics. `target_motor_cmd` values are only for the testing purposes___
![image](https://user-images.githubusercontent.com/7955120/95956073-bf936f00-0e38-11eb-9484-f8c777786985.png)
```bash
# Commands sent through
----- 1st command ------
[INFO] [1602658772.604579, 17.700000]: header: 
  seq: 0
  stamp: 
    secs: 17
    nsecs: 700000000
  frame_id: ''
pitch_cmd_type: 2
target_pitch_value: 0.4
motor_cmd_type: 2
target_motor_cmd: 2
rudder_control_mode: 0
target_heading: 0.0
rudder_angle: 0
target_rudder_angle: 0.0
target_pumped_volume: 0.0

----- 2nd command ------
[INFO] [1602658777.607426, 22.700000]: header: 
  seq: 0
  stamp: 
    secs: 22
    nsecs: 700000000
  frame_id: ''
pitch_cmd_type: 2
target_pitch_value: 0.6
motor_cmd_type: 2
target_motor_cmd: 4
rudder_control_mode: 0
target_heading: 0.0
rudder_angle: 0
target_rudder_angle: 0.0
target_pumped_volume: 0.0

----- 3rd command ------
[INFO] [1602658782.615358, 27.700000]: header: 
  seq: 0
  stamp: 
    secs: 27
    nsecs: 700000000
  frame_id: ''
pitch_cmd_type: 2
target_pitch_value: 0.2
motor_cmd_type: 2
target_motor_cmd: 4
rudder_control_mode: 0
target_heading: 0.0
rudder_angle: 0
target_rudder_angle: 0.0
target_pumped_volume: 0.0

----- 4th command ------
[INFO] [1602658787.623409, 32.700000]: header: 
  seq: 0
  stamp: 
    secs: 32
    nsecs: 700000000
  frame_id: ''
pitch_cmd_type: 2
target_pitch_value: 0.0
motor_cmd_type: 2
target_motor_cmd: 2
rudder_control_mode: 0
target_heading: 0.0
rudder_angle: 0
target_rudder_angle: 0.0
target_pumped_volume: 0.0

----- 5th command ------
[INFO] [1602658792.631357, 37.700000]: header: 
  seq: 0
  stamp: 
    secs: 37
    nsecs: 700000000
  frame_id: ''
pitch_cmd_type: 2
target_pitch_value: -0.2
motor_cmd_type: 2
target_motor_cmd: 4
rudder_control_mode: 0
target_heading: 0.0
rudder_angle: 0
target_rudder_angle: 0.0
target_pumped_volume: 0.0

----- 5th command ------
[INFO] [1602658797.639443, 42.700000]: header: 
  seq: 0
  stamp: 
    secs: 42
    nsecs: 700000000
  frame_id: ''
pitch_cmd_type: 2
target_pitch_value: 0.0
motor_cmd_type: 2
target_motor_cmd: 1
rudder_control_mode: 0
target_heading: 0.0
rudder_angle: 0
target_rudder_angle: 0.0
target_pumped_volume: 0.0
```

woensug-choi committed:
initiation of the new branch

woensug-choi committed:
basic set-up first init

woensug-choi committed:
initial templates ready for commanding with frl_vehicle_msgs

woensug-choi committed:
all working independently, however, pitch and rudder control stop the vehicle. (subscribing function needs check)

woensug-choi committed:
Commanding function ready

daewok commented:
Can you prioritize a basic status publisher over lat/lon? I think having depth be the only value sent is fine for the first iteration. The "hello world" test for integrating with the simulated glider flight computer is going to be mission ini0.mi, a stock mission that just goes down several meters before surfacing again. Without the depth being reported, the controller won't know when to trigger the inflection back upward.

woensug-choi committed:
UwGliderStatus puglisher for depth and RPY added

woensug-choi commented:
> Can you prioritize a basic status publisher over lat/lon? I think having depth be the only value sent is fine for the first iteration. The "hello world" test for integrating with the simulated glider flight computer is going to be mission ini0.mi, a stock mission that just goes down several meters before surfacing again. Without the depth being reported, the controller won't know when to trigger the inflection back upward.

Thank you for re-prioritizing. Moved lat/lon to after sensors (lat/lon from GPS sensor I suppose). UwGliderStatus publishing function is added!
topics are
- commands : `/glider_hybrid_whoi/direct_kinematics/UwGliderCommand`
- status: `/glider_hybrid_whoi/direct_kinematics/UwGliderStatus`

Currently, only `pitch_cmd_type=2` and `motor_cmd_type=2` is available. Would that be enough for our starter?

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of github.com:Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
Added DVL sensor and new worlds for altitude status

woensug-choi commented:
#### Updates
- Added sensors including DVL, Camera, GPS, IMU (using UUV simulator's)
- Added subscriber of DVL altitude and included it in UwGliderStatus
- New world and test python script considering seafloor

#### New commands
New test commands would be as follows (all in separate terminal windows)
```bash
roslaunch glider_hybrid_whoi_gazebo seafloor_underwater_world.launch paused:=false
roslaunch glider_hybrid_whoi_gazebo start_demo_teleop_kinematics.launch joy_id:=0
rosrun direct_kinematics_ros_plugins test_directKinematics.py
```

#### Results
![Capture](https://user-images.githubusercontent.com/7955120/96214368-3e61e680-0fb6-11eb-9227-8ebc0195445e.PNG)

woensug-choi committed:
connected gps sensor data

woensug-choi committed:
fixed minor sign errors

woensug-choi committed:
initial buoyancy engine added (cCOB=COG)

woensug-choi committed:
pumped volume unit fix

woensug-choi committed:
added commands with target_pumped_volume

woensug-choi committed:
visual effects for propeller rotation added

woensug-choi committed:
dynamics added

woensug-choi committed:
dynamics on if pitch control type is battery position

woensug-choi commented:
### Current development status and notes
The new kinematics plugin is almost ready to review and merge! It may require polishing later. But I believe it will serve well for the tests with the control system and backseat driver. It includes dynamics functionality when the battery (=slinding mass, internal mass) position is controlled. Otherwise, kinematics manipulation using the generic Gazebo-ROS service function `/gazebo/set_model_state` is used on prompt instantly.

#### Pitch control
| Battery position | Target once | Target servo |
| ------------- | ------------- | ------------- |
| :heavy_check_mark: w Dynamics*  | :heavy_check_mark:  | :heavy_multiplication_x: |
- Dynamics that was verified with the MATLAB code is migrated. 

#### Motor control
| Voltage | Power |
| ------------- | ------------- | 
| :heavy_multiplication_x:  | :heavy_check_mark: |
- Power command (in range 1-9) is applied as a force in [N] linear twist in body frame.
- Asked Pete to provide the final interpolation function of our new thruster.
- Propeller rotation is visualized using the command value as rotation speed.

#### Rudder control
| Heading | Angle (Center, Port, Stbd) | Angle (Direct) |
| ------------- | ------------- | ------------- | 
| :heavy_multiplication_x:  | :heavy_multiplication_x: | :heavy_multiplication_x: |
- The size of the port and starboard angle is set to PI/3.

#### Buoyancy engine control
| Kinematics | Dynamics |
| ------------- | ------------- |
| :heavy_check_mark:  | :heavy_check_mark: |

#### Glider status
| Lat/Lon | Roll-Pitch-Yaw | Heading | Depth | Altitude | Power | Rudder angle | Battery position | Pumped Volume |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| :heavy_check_mark:*  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:* | :heavy_multiplication_x: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
- Lat/Lon is acquired from the GPS sensor and Altitude from the DVL sensor.

#### How-to
- Running the simulator
```bash
roslaunch glider_hybrid_whoi_gazebo seafloor_underwater_world.launch paused:=true
roslaunch glider_hybrid_whoi_gazebo start_demo_teleop_kinematics.launch joy_id:=0
```
- Sending commands to the glider
```bash
# Using test python script
rosrun direct_kinematics_ros_plugins test_directKinematics.py
# directly to the ros node (non defined values will be assumed as zero)
rostopic pub /glider_hybrid_whoi/direct_kinematics/UwGliderCommand frl_vehicle_msgs/UwGliderCommand '{pitch_cmd_type: 2, target_pitch_value: 0.5, motor_cmd_type: 2, target_motor_cmd: 1, target_pumped_volume: 0.0001 }'
```
- Glider status
```bash
rostopic echo /glider_hybrid_whoi/direct_kinematics/UwGliderStatus
```

#### Other notes
- Currently, the 3D Pose, DVL, IMU, Camera, GPS, and its battery consumer macros (from UUV simulator) are attached.
- Dynamics functionality is capable of 6-DOF, but it is verified only in 3-DOF.
- The simulation assumes zero gravity in the world configuration (Neutral buoyancy function did not work at the UUV simulator). 
The current case set-up does have realistic parameters for inertial and drag properties, but not the gliders' exact values.
- The sensors attached in the current case are inherited from ECA A9, which is huge for our glider.
- The plugin contains both kinematics and dynamics control. New name the plugin..?
- The plugin is written in single .hh and .cc file set. I think it's ok for now.

### Previous To-dos
- [x] Add publisher for glider status with frl_vehicle_msgs
- [x] Fix propellers not showing errors. (Unable to get propeller links by this-model()->GetLinks())
- [x] Add sensors and lat/lon/altitude status publishing
- [x] Add buoyancy pump control function
- [x] Add the interpolated function of our new thruster
   - Pete will provide the final function for our new thruster
- [x] Investigate precision of command send time and glider response time delays
   - Current time delay between command and status is 0.1 [s] as defined at `.world` file as `real_time_update_rate=10`
- [x] Add joint control function for manipulation of rudders/propellers visually
- [x] Add minimalistic dynamics characteristics
- [x] Diagrams for plugin wiki & Record a video of the feature with realistic case set-ups
   - This will be done after connecting with Eric's part and set-up based on the mission files.

#### Future to-dos
The following to-dos are to be expected. After merging of this PR, it will be moved to Issues.
If anything needs to be added to the list, please let me know. **I am not expecting to do this anytime soon. Polishing the plugin will be my priority.**
- [x] Add submergence to detect ocean surface

daewok commented:
I've been working on tying this into my stuff and I'm having some issues.

I've been testing with the following command: `{pitch_cmd_type: 2, target_pitch_value: 0.4538, target_pumped_volume: -260.0}`.

First, it appears that a command needs to received within a fairly small time window after the simulation has been unpaused (on the order of 2 seconds). If a command comes while the simulation is still paused or after that window, the glider seems to start drifting and once the command is received, it immediately pitches nose straight down and floats upwards. I can't reliably hit this time window with the simulated glider control system, so I've been resorting to manually sending messages.

Second, if I manage to hit this time window, it appears the glider does not move in the X-Y plane unless the thruster is on as well. The stock checkout dives (our hello world) do not use the thruster and in practice we don't always run the thruster so as to save power.

Third, it seems there is nothing preventing the glider from leaving the water. I'm guessing this is what you wanted to fix with the "Add submergence to detect ocean surface" TODO. Is it possible to move up the priority of that? I *could* add some checks to try and prevent it on my side, but since I'm running on a 0.25 Hz control loop and not controlling z directly it seems like the glider would routinely end up floating out of the water at least a little bit.

UPDATE: This behavior I've observed seems stochastic... Sometimes it doesn't go crazy if I leave it sit a long time before sending a command. But I've yet to determine when the triggering condition is.

woensug-choi committed:
Iter 1, Debug trial 1

woensug-choi commented:
@daewok First thing came up to my mind was that the plugin has turned on 'dynamics' function even when the pitch_control isn't battery control.

Added a commit on PR not to. Could you give it a try as a short feedback?

I shall make a diagram and a video capture from my side soon.

daewok commented:
The latest commit worked fixed the glider floating like a balloon for me 3 out of 3 times. I'll hammer on it more over the next few days and let you know if I see anything wonky.

It seems there's still the issue of there being no X-Y motion if the thruster is off. Is that a reasonable thing to fix?

Last, I haven't looked into it too in depth (ha), but the rate of descent for -260cc pumped volume feels way too fast. I'm getting something like 2.61 m/s. Is that derived from some physical model currently or just from past dive data? I'll try to dig up some data to see if I'm misremembering/what the typical descent rates are.

woensug-choi committed:
buoyancy fixed, thruster changed to using AddRelativeForee, Rudder control excluded

woensug-choi commented:
> The latest commit worked fixed the glider floating like a balloon for me 3 out of 3 times. I'll hammer on it more over the next few days and let you know if I see anything wonky.
> 
> It seems there's still the issue of there being no X-Y motion if the thruster is off. Is that a reasonable thing to fix?
> 
> Last, I haven't looked into it too in depth (ha), but the rate of descent for -260cc pumped volume feels way too fast. I'm getting something like 2.61 m/s. Is that derived from some physical model currently or just from past dive data? I'll try to dig up some data to see if I'm misremembering/what the typical descent rates are.

Glad to hear that my hunch was correct. I've updated with a commit after finding something strange behavior with rudder controls. I have excluded it for now. And some other changes including the fix for the buoyancy force. I am not sure why the previous methods give a large force. Made changes to do the same thing with other methods. It should be fine now.

Motion in X-Y direction without the thruster brings up a rather big question. Kinematics manipulation does not include dynamics. Are there any mission files doing a dive including thruster usage? Manipulating only pitch and buoyancy engine to move in X-Y direction requires the dynamics of the glider. Currently, the dynamics are migrated to the battery position command only. 

You may try with `rostopic pub -1 /glider_hybrid_whoi/direct_kinematics/UwGliderCommand frl_vehicle_msgs/UwGliderCommand '{pitch_cmd_type: 1, target_pitch_value: 0.002, target_pumped_volume: -250}'`. Unfortunately, it is not fully usable since it diverges after some time. Maybe it could stay converged with realistic inputs. The cause of this is yet to be found (maybe can be solved by changing the world physic solver in the gazebo?) 
(Update. it's been temporarily fixed with linear force to Y direction and torque to Yaw and Roll set to zero values).

woensug-choi committed:
buoyancy sign fixed for dynamics

woensug-choi committed:
Temporarily restricted forces in dynamics to 3 dimension only for stability

woensug-choi commented:
Had to remove the rudder control function entirely after finding unknown behavior. Does anyone know the reason for this? 
Help please! @bsb808 @mabelzhang 
#### Test Code
```cpp
  ignition::math::Quaterniond orientation(this->modelState.pose.orientation.x,
                                          this->modelState.pose.orientation.y,
                                          this->modelState.pose.orientation.z,
                                          this->modelState.pose.orientation.w);
  ignition::math::Vector3d orientation_euler = orientation.Euler();
  gzmsg << "=====================" << std::endl;
  gzmsg << orientation_euler << std::endl;
  ignition::math::Quaterniond orientation_quat;
  orientation_quat.Euler(orientation_euler.X(),
                         orientation_euler.Y(),
                         orientation_euler.Z());
  
  gazebo_msgs::SetModelState cmd_msg;
  cmd_msg.request.model_state = this->modelState;
  cmd_msg.request.model_state.pose.orientation.x = orientation_quat.X();
  cmd_msg.request.model_state.pose.orientation.y = orientation_quat.Y();
  cmd_msg.request.model_state.pose.orientation.z = orientation_quat.Z();
  cmd_msg.request.model_state.pose.orientation.w = orientation_quat.W();
```
#### Results
```bash
=====================
0 -0 3.14159
=====================
3.14159 0 -3.14159
=====================
-3.14159 -0 0
=====================
0 0 -0
=====================
-0 -0 3.14159
```

As the code says, It doesn't change anything. Receives `this->modelState` updated at every gazebo frame and receives it and set it with `Set_gazebo_state`. But the values are changing with some kind of a pattern.

mabelzhang commented:
I guess there are two places that can go wrong.

Are the two quaternions `orientation` and `orientation_quat` the same?
I'm wondering if there's a convention difference in converting to and from the Euler angles. Though theoretically I would guess `ignition::math::Quaternion` should be using the same convention in its functions.

Does this happen if you use `orientation` to `SetModelState`, as opposed to `orientation_quat`?
That would isolate the problem to something in actually setting the state and not the quaternion itself.

daewok commented:
I've gotten some more testing done on the hello world and it's almost there.

The hello world triggers the dynamics functions because it controls the battery pack position directly, so that means we even have forward motion (yay!).

However, the glider can quite easily float out of the water when in dynamics mode which, while probably not a show stopper for the hello world, is definitely weird to see.

Second, the plugin can switch out of dynamics mode before it's reached steady state. If I send `pitch_cmd_type: 1, target_pitch_value: 0.02` and then in the next control cycle send `pitch_cmd_type: 0`, the vehicle just instantly stops pitching and never reaches the target pitch value.

woensug-choi commented:
> I guess there are two places that can go wrong.
> 
> Are the two quaternions `orientation` and `orientation_quat` the same?
> I'm wondering if there's a convention difference in converting to and from the Euler angles. Though theoretically I would guess `ignition::math::Quaternion` should be using the same convention in its functions.
> 
> Does this happen if you use `orientation` to `SetModelState`, as opposed to `orientation_quat`?
> That would isolate the problem to something in actually setting the state and not the quaternion itself.
```cpp
 ignition::math::Quaterniond orientation(this->modelState.pose.orientation.x,
                                          this->modelState.pose.orientation.y,
                                          this->modelState.pose.orientation.z,
                                          this->modelState.pose.orientation.w);
  ignition::math::Vector3d orientation_euler = orientation.Euler();
  gzmsg << "=====================" << std::endl;
  gzmsg << orientation_euler << std::endl;  
  gazebo_msgs::SetModelState cmd_msg;
  cmd_msg.request.model_state = this->modelState;
  cmd_msg.request.model_state.pose.orientation.x = orientation.X();
  cmd_msg.request.model_state.pose.orientation.y = orientation.Y();
  cmd_msg.request.model_state.pose.orientation.z = orientation.Z();
  cmd_msg.request.model_state.pose.orientation.w = orientation.W();

  // this->commandPublisher["Model"].call(cmd_msg);
```
With the above code, without publishing the command through set_model_state but just printing what's been received using get_model_state.
- Run
- Empty custom command sent (doesn't change the model, since publish function is commented at as above)
```bash
=====================
0 -0 3.14159
```
- Set model state command using another terminal
```
# command sent using generic gazebo command
rostopic pub -1 /gazebo/set_model_state gazebo_msgs/ModelState "{model_name: 'glider_hybrid_whoi', pose: {orientation: {x: 0.707107, y: 0.707107, z: 0.0, w:0.0}}}"
```
- Empty custom command sent (doesn't change the model, since publish function is commented at as above)
```bash
=====================
1.5708 -0 0
```
which corresponds to `orientation: {x: 0.707107, y: 0, z:0, w: 0.707107}` 👎 It uses the numbers given, but in different axis notation or something. I can't imagine why and how...

Something I don't fully understand is in the plugin. 'NED coordinate frame transformers' [link](https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/5f017d6f355657c82697c9b0ccf6b7002d08fd52/direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc#L196) [link2](https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/5f017d6f355657c82697c9b0ccf6b7002d08fd52/direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc#L322) Maybe these has to do with this phenomenon.

woensug-choi commented:
> I've gotten some more testing done on the hello world and it's almost there.
> 
> The hello world triggers the dynamics functions because it controls the battery pack position directly, so that means we even have forward motion (yay!).
> 
> However, the glider can quite easily float out of the water when in dynamics mode which, while probably not a show stopper for the hello world, is definitely weird to see.
> 
> Second, the plugin can switch out of dynamics mode before it's reached steady state. If I send `pitch_cmd_type: 1, target_pitch_value: 0.02` and then in the next control cycle send `pitch_cmd_type: 0`, the vehicle just instantly stops pitching and never reaches the target pitch value.

Glad to hear that something is finally working! (yay! 👍 ) I guess that the kinematics are not useful at the moment (It can't move in X-Y directions). It's lucky that I had dynamics that can be migrated. Let me know if you find any other strange behaviors that are critical.

I've added modifications so that `pitch_cmd_type: 0` would not change the regime of dynamics and kinematics (It was in the kinematics realm). Pitch command types define the regime of dynamics and kinematics. It's using different function sets. `pitch_cmd_type: 1` tells the plugin to use dynamics functions (setting `this->dynamics = True` and keeps it that way unless type 2 or 3 is being used). A mixture of dynamics and kinematics will be introduced after PID controls are added.

Free surface detection added. Whenever the vehicle hits the free surface from underwater, the force, velocity, and accelerations that were being applied to the vehicle are stripped off. It's not fully physical (no wave motion; no torque according to partial submergence), but I believe it will do enough to do simple missions. (*NOTE: our frl_msgs `target_pumped_volume` does not have the NONE command type. Setting it to zero literally means zero position of the pump pistons. Which causes problems when only pitch/thruster commands are sent (only pitch command=zero value for target_pump=zero buoyancy force). I've manually set zero value to be the NONE commanding type for the buoyancy engine. Use minimal values (e.g. 0.001) to turn the buoyancy engine off. On the surface, set it minus values (e.g. -0.001) to have the plugin recognize that the buoyancy engine is off and not pushing the vehicle to the surface. Otherwise, it will keep the vehicle on the surface by striping any commanding forces off.

woensug-choi committed:
pitch_cmd_type: 0 option reconsidered

woensug-choi committed:
added a function for our thruster (voltage input needs investigation)

woensug-choi committed:
multiply 2.0 for thruster power value (dual thrusters)

woensug-choi committed:
free surface detection added

woensug-choi commented:
![image](https://user-images.githubusercontent.com/7955120/97170070-6deecb00-17ce-11eb-9a20-cc5a03093e3a.png)

daewok commented:
https://youtu.be/GXGha0PuF3s

daewok committed:
Improve interface diagram

Add slocum_glider_sim_driver and slocum_glider_sim_console nodes. Make clear
that the "backseat driver" interacts with the sim driver via its extctl
interface. Provide link to where these nodes are located.

woensug-choi committed:
Updating plugin diagram

woensug-choi committed:
Usable kinematics

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of github.com:Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
Usable Kinematics functionality

daewok commented:
The kinematics mode is looking pretty good now. Just two things I noticed in testing:

1. It looks like the requested heading is being interpreted in ENU instead of NED. I'm doing a box pattern of North->East->South->West. When I run it, the glider starts facing what I assume is East and makes 90 degree turns to the left.

2. It looks like something is broken with surface detection. Every time the glider clips the surface it imparts some angular velocity to the glider which is never cancelled out (since the kinematics path never explicitly sets twist.angular). Additionally, if I explicitly try to surface, the glider just keeps on ascending in the air.

I can send bag files and/or videos if it would help.

daewok commented on direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc:
```diff
@@ -0,0 +1,1245 @@
+  cmd_msg.request.model_state.pose.orientation.y = target_orientation.Y();
+  cmd_msg.request.model_state.pose.orientation.z = target_orientation.Z();
+  cmd_msg.request.model_state.pose.orientation.w = target_orientation.W();
+  this->commandPublisher["Model"].call(cmd_msg);
```
Why is the service called here and then again after the following if statement?

woensug-choi replied:
You are correct. It's redundent. I'll erase it.

daewok commented on direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc:
```diff
@@ -0,0 +1,1245 @@
+  int _motor_cmd_type = _msg->motor_cmd_type;
+  double _motor_cmd_value = _msg->target_motor_cmd;
+  // 1 : voltage control, 2: power control
+  if (_motor_cmd_type == 1 || 2)
```
You should really refer to these constants by name instead of value. In Python we could do `UwGliderCommand.MOTOR_CMD_VOLTAGE`. I would assume the C++ bindings have something similar.

woensug-choi replied:
I don't fully understand this comment. Would it cause problem? You mean **int** _motor_cmd_type should be something like **char**?

daewok replied:
Ha, my original comment was about a style issue, but looking at this line again I see there's an actual bug.

Style issue: There's nothing technically wrong with something like `_motor_cmd_type == 1`, but it's both not obvious what `1` refers to and it has to be kept in sync with the message definition if the value of voltage control ever changes. Something like `_motor_cmd_type == _msg->MOTOR_CMD_VOLTAGE` (no idea if that's the right syntax) is more self documenting and robust to minor changes in the message file.

Bug: This condition always evaluates to true since 2 is always truthy. You probably meant `(_motor_cmd_type == 1 || _motor_cmd_type == 2)`

woensug-choi committed:
surface detection for kinematics fixed

woensug-choi commented:
> The kinematics mode is looking pretty good now. Just two things I noticed in testing:
> 
>     1. It looks like the requested heading is being interpreted in ENU instead of NED. I'm doing a box pattern of North->East->South->West. When I run it, the glider starts facing what I assume is East and makes 90 degree turns to the left.
> 
>     2. It looks like something is broken with surface detection. Every time the glider clips the surface it imparts some angular velocity to the glider which is never cancelled out (since the kinematics path never explicitly sets twist.angular). Additionally, if I explicitly try to surface, the glider just keeps on ascending in the air.
> 
> 
> I can send bag files and/or videos if it would help.

Glad to hear it works and meet your needs! :+1: And thank you for reviewing the source code!

1. Confirmed that it's in ENU frame can change to NED giving it a negative sign to read msg. Can you explain more about `When I run it, the glider starts facing what I assume is East and makes 90 degree turns to the left.`? I will wait to change it to NED until I fully understand your situation here.

2. I've fixed surface detection and made a commit.
tested with following commands
```bash
# Toward surface
rostopic pub /glider_hybrid_whoi/direct_kinematics/UwGlidCommand frl_vehicle_msgs/UwGliderCommand '{motor_cmd_type: 1, target_motor_cmd: 2, pitch_cmd_type: 2, target_pitch_value: -0.4, rudder_control_mode: 1, target_heading: 0.8}'
# Back to underwater
rostopic pub /glider_hybrid_whoi/direct_kinematics/UwGlidCommand frl_vehicle_msgs/UwGliderCommand '{motor_cmd_type: 1, target_motor_cmd: 2, pitch_cmd_type: 2, target_pitch_value: 0.4, rudder_control_mode: 1, target_heading: -0.8}'
```

daewok commented:
Surface detection is partially fixed. I no longer get the angular velocities, but it's not constrained to stay in the water. Perhaps it's related to the buoyancy engine? I noticed your test didn't set it. This video is from right after the simulated control system boots. The first thing it does is make itself safe by setting the buoyancy engine to max buoyancy (but there are plenty of other times we go to max buoyancy and keep it there to stay on the surface).

https://www.youtube.com/watch?v=mLXW_9MOKX8

The glider starts at .75m depth and the first message is: `{rudder_angle: 4, target_pumped_volume: 226.0}` (the rudder angle is just an artifact, I assume it has no effect since `rudder_control_mode` is 0).

>Confirmed that it's in ENU frame can change to NED giving it a negative sign to read msg. Can you explain more about `When I run it, the glider starts facing what I assume is East and makes 90 degree turns to the left.`? I will wait to change it to NED until I fully understand your situation here.

I was just trying to say that my current test mission drives the glider in a box going North, East, South, and West. When I send those commands the glider seems to start facing East and then turns to the North, West, and South.

So the fix is probably a combo sign flip and rotation to make North be zero heading.

> Glad to hear it works and meet your needs! +1 And thank you for reviewing the source code!

By no means was my review thorough! I just commented on the things that stuck out to me on first glance. You'll definitely want like @bsb808 to review as well.

daewok commented on direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc:
```diff
@@ -0,0 +1,1247 @@
+  double target_pitch = this->prev_pitch;
+  int _pitch_cmd_type = _msg->pitch_cmd_type;
+  // 1 : battery position, 2: target once , 3: target servo
+  if (_pitch_cmd_type == 1)  // battery position
```
Consider using `switch` here instead of a bunch of `if/else if/else`. I find it easier to read and it really shines in cases like this where you're dispatching on the value of a single variable,

woensug-choi replied:
:+1:

woensug-choi committed:
Surface detection fix, heading angle fix, switch statement conversion

woensug-choi commented:
@daewok Both problems (surface detection and heading angle) is now fixed.
- Surface detection had its detecting condition with positive buoyancy force of dynamics. Added additional 'or' condition for positive buoyancy command msg too.
- Heading angle is fixed with adding pi/2 and minus sign to command msg.

daewok committed:
In the kinematic case, servoing on pitch is equivalent to setting it once

daewok commented:
Seems to be working. I'll let you know if I run into any more problems as I really dig in.

https://youtu.be/KXYxdnajWaU

woensug-choi commented:
If we get to the intermediate conclusion that we have agreeable kinematics functionality working, I will strip off unnecessary portions of this PR and make a new clean PR for it with a Wiki description page. Of course, some parameter adjustments may be brought at issues after merging of the new PR.

woensug-choi committed:
add ci check

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of https://github.com/Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
fix ci check

woensug-choi committed:
add code_check script

woensug-choi committed:
remove ci check, uuv_simulator inhereted code cause tons of problems

daewok commented:
Sounds like a plan to me. I haven't had a chance to really test as much as I wanted, but I'll try to do so by next week's meeting.

Also, I believe we talked about Dockerizing everything for easy deployment? Have you given that much thought for this repo?

woensug-choi commented:
Dockerizing it would not be difficult. I could make an Dockerfile and short tutorials about how to deploy the docker container.

woensug-choi commented:
@daewok How was your holiday? Hope you had some good time with your family! 
Hope nothing critical problemes during plugin tests. Tried to add some static code checkings to this PR, but codes inherited from the UUV simulator cause too many problems.

daewok commented:
No other show stoppers yet. But it would be nice to have a way to distinguish "dead reckoned" lat/lng from GPS lat/lng. I put one possible way of doing that in https://github.com/Field-Robotics-Lab/frl_msgs/pull/8

daewok commented:
For reference, my Dockerfile is at: https://gitlab.com/mit-mers/ros/slocum_glider/-/blob/master/Dockerfile

The image name for the master branch is `registry.gitlab.com/mit-mers/ros/slocum_glider:master`. The `latest` tag will be used once I cut a release (I'll probably do that after I make sure our images play well together).

Once you get this Dockerized, we can pretty easily put together a docker-compose file.

woensug-choi commented:
#### To-dos for First half of December
- [x] Strip-off dynamics part and make new Kinematics PR (including Docker)
- [x] Add nav_sat_fix to glider status (https://github.com/Field-Robotics-Lab/frl_msgs/pull/8)
- [ ] Standalone case-set PR with Buzzard Bay bathymetry & Stratified ocean current (with Kolumbo data)

woensug-choi committed:
add sat_nav_fix.msg

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of github.com:Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi commented:
New PR made at https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/pull/16
